File: facerec_online.py
#!/usr/bin/env python
#
# an online / retrainable facerec version using the webcam.
#
# since the face reco code was taken from opencv/samples/python2/face_rec.py
# we won't spare you the licence ;)
#
# Software License Agreement (BSD License)
#
# Copyright (c) 2012, Philipp Wagner <bytefish[at]gmx[dot]de>.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of the author nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

# ------------------------------------------------------------------------------------------------
# Note:
# When using the FaceRecognizer interface in combination with Python, please stick to Python 2.
# Some underlying scripts like create_csv will not work in other versions, like Python 3.
# ------------------------------------------------------------------------------------------------

import os
import sys
import cv2
import numpy as np

import random
rand = random.Random()

# added a names list(z)
def read_images(path, sz=None):
    """Reads the images in a given folder, resizes images on the fly if size is given.

    Args:
        path: Path to a folder with subfolders representing the subjects (persons).
        sz: A tuple with the size Resizes

    Returns:
        A list [X,y]

            X: The images, which is a Python list of numpy arrays.
            y: The corresponding labels (the unique number of the subject, person) in a Python list.
    """
    c = 0
    X,y,z = [], [], []
    for dirname, dirnames, filenames in os.walk(path):
        for subdirname in dirnames:
            subject_path = os.path.join(dirname, subdirname)
            for filename in os.listdir(subject_path):
                try:
                    im = cv2.imread(os.path.join(subject_path, filename), cv2.IMREAD_GRAYSCALE)
                    if (len(im)==0):
                        continue # not an image                        
                    # resize to given size (if given)
                    if (sz is not None):
                        im = cv2.resize(im, sz)
                    X.append(np.asarray(im, dtype=np.uint8))
                    y.append(c)
                except IOError, (errno, strerror):
                    print "I/O error({0}): {1}".format(errno, strerror)
                except:
                    print "Unexpected error:", sys.exc_info()[0]
                    raise
            c = c+1
            z.append(subdirname)
    return [X,y,z]



# reload the images & retrain the model.
# note, that lbp would give you the possibility of 
# just updating the model with additonal data instead.
def retrain( imgpath, model,sz ) :
    # read in the image data. This must be a valid path!
    X,y,names = read_images(imgpath,sz)
    if len(X) == 0:
        print "image path borked", imgpath
        return [[],[],[]]
    # Learn the model. Remember our function returns Python lists,
    # so we use np.asarray to turn them into NumPy lists to make
    # the OpenCV wrapper happy:
    # Also convert labels to 32bit integers. This is a workaround for 64bit machines,
    model.train(np.asarray(X), np.asarray(y, dtype=np.int32))
    return [X,y,names]



if __name__ == "__main__":
    # You'll need the path to your image folder, also we need to find 
    # a haar/lbpcascade for detecting faces, e.g. opencv\data\haarcascades\haarcascade_frontalface_alt2.xml
    if len(sys.argv) < 3:
        print "USAGE: facerec_online.py </path/to/images> <path/to/cascadefile>"
        sys.exit()
        
    print "  press 'esc' to quit"
    print "  press 'a' to append a new face to the database"
    print "      (you'll be prompted for a name on the console)"
    print "  press 't' to retrain the model (if you appended faces there)"
    
    # create the img folder, if nessecary
    imgdir = sys.argv[1]
    try:
        os.mkdir(imgdir)
    except:
        pass # dir already existed

    # default face size, all faces in the db need to be the same.
    face_size=(90,90)
    
    # open the webcam
    cam = cv2.VideoCapture(0)
    if ( not cam.isOpened() ):
         print "no cam!"
         sys.exit()      
    print "cam: ok."         
    
    # load the cascadefile:
    cascade = cv2.CascadeClassifier(sys.argv[2])
    if ( cascade.empty() ):
         print "no cascade!"
         sys.exit()         
    print "cascade:",sys.argv[2]
    
    # Create the model. We are going to use the default
    # parameters for this simple example, please read the documentation
    #model = cv2.createEigenFaceRecognizer()
    #model = cv2.createFisherFaceRecognizer()
    model = cv2.createLBPHFaceRecognizer()
    
    # train it from faces in the imgdir:
    images,labels,names = retrain(imgdir,model,face_size)
    print "trained:",len(images),"images",len(names),"persons"
    
    while True:
        ret, img = cam.read()
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
        gray = cv2.equalizeHist(gray)
        # try to detect a face in the img:
        rects = cascade.detectMultiScale(gray, scaleFactor=1.3, minNeighbors=4, minSize=(30, 30), flags = cv2.CASCADE_SCALE_IMAGE)       
        
        # roi will keep the cropped face image ( if there was one )
        roi = None
        for x, y, w, h in rects:
            # crop & resize it 
            roi = cv2.resize( gray[y:y+h, x:x+h], face_size )
            # give some visual feedback for the cascade detection
            cv2.rectangle(img, (x,y),(x+w,y+h), (255, 0, 0))
            if len(images)>0:
                # model.predict is going to return the predicted label and
                # the associated confidence:
                [p_label, p_confidence] = model.predict(np.asarray(roi))
                cv2.putText( img, "%s %.2f" % (names[p_label], p_confidence),(x+10,y+20), cv2.FONT_HERSHEY_PLAIN,1.3, (0,200,0))
            break # use only 1st detected

        cv2.imshow('facedetect', img)

        k = cv2.waitKey(5) & 0xFF
        
        # bailout on 'esc'
        if k == 27: break
       
        # 'a' pressed, add person to the database
        if (k == 97) and (roi!=None): 
            print "please input the name: "
            name = sys.stdin.readline().strip('\r').strip('\n')
            # make a folder for that person:
            dirname = os.path.join(imgdir,name)
            try:
                os.mkdir(dirname)
            except:
                pass # dir already existed
            # save image
            path=os.path.join(dirname,"%d.png" %(rand.uniform(0,10000)))
            print "added:",path
            cv2.imwrite(path,roi)
        
        # if enough new data was collected, retrain the model
        if (k == 116): # 't' pressed
            images,labels,names = retrain(imgdir,model,face_size)
            print "trained:",len(images),"images",len(names),"persons"
                         

File: facerec_online.cpp
//
// realtime online demo.
// instead of using/parsing a csv file,
// we save each person to a seperate directory,
// and later parse the glob() output.
//
// YOU NEED AN EMPTY "persons" FOLDER HERE!
// (train images will get saves there,
//   on startup, the app will try to train from those images)
//
// use : online [capture id or path] [img_path] [cascade_path]
//
//

#include <opencv2/opencv.hpp>
#include <opencv2/face.hpp>

#include <iostream>
#include <fstream>
#include <sstream>
#include <map>

using namespace std;
using namespace cv;

#ifdef _WIN32
 const char SEP = '\\';
#else
 const char SEP = '/';
#endif


//
// 3 state online face db
//
enum
{
    NEUTRAL = 0,
    CAPTURE = 1,
    PREDICT = 2
};

//
// default face size, [80..160]
//
enum
{
    FIXED_FACE = 90
};



class FaceRec
{
    Ptr<face::FaceRecognizer>  rec;

public:
    FaceRec()
        : rec(face::LBPHFaceRecognizer::create())
    {}

    Mat preproc(const Mat &img) {
        Mat gray, rez;
        if (img.channels()>1)
            cvtColor(img, gray, COLOR_BGR2GRAY);
        else gray = img;
        const Size cs(FIXED_FACE, FIXED_FACE);
        if (gray.size() != cs) {
            resize(gray, rez, cs);
            return rez;
        }
        return gray;
    }
    int train(const String &imgdir) {
        string last_n("");
        int label(-1);
        map<int,int> classes;
        vector<Mat> features;
        vector<int> labels;

        vector<String> vec;
        glob(imgdir,vec,true);
        if ( vec.empty())
            return 0;

        for (size_t i=0; i<vec.size(); i++) {
            // extract name from filepath:
            String v = vec[i];
            int r1 = v.find_last_of(SEP);
            String v2 = v.substr(0,r1);
            int r2 = v2.find_last_of(SEP);
            String n = v2.substr(r2+1);
            if (n!=last_n) {
                last_n=n;
                label++;
            }
            rec->setLabelInfo(label, n);

            // process img & add to trainset:
            Mat img=imread(vec[i],0);
            features.push_back(preproc(img));
            labels.push_back(label);
            if (classes.find(label)==classes.end())
                classes[label] = 1;
            else
                classes[label] ++;
        }
        rec->train(features, labels);
        for (auto it=classes.begin(); it!=classes.end(); ++it) {
            String name = rec->getLabelInfo(it->first);
            cerr << format("%25s %4d", name.c_str(), it->second) << " images." << endl;
        }
        return classes.size();
    }

    String predict(const Mat & img) {
        int id=-1;
        double dist=-1;
        rec->predict(preproc(img), id, dist);
        if (id < 0)
            return "";

        float conf = 1.0f - sqrt(dist*dist)/250;
        String name = rec->getLabelInfo(id);
        return format("%s : %2.3f", (dist<90?name.c_str():"unknown"), dist);
    }

    bool load(const String &fn) {
        rec = Algorithm::load<face::LBPHFaceRecognizer>(fn);
        return 1;
    }
    bool save(const String &fn) {
        rec->save(fn);
        return 1;
    }
};

int main(int argc, const char *argv[])
{
    theRNG().state = getTickCount();

    cerr << "press 'c' to record new persons," << endl;
    cerr << "      space, to stop recording. (then input a name)." << endl;
    cerr << "      'p' to predict," << endl;
    cerr << "      'n' for neutral," << endl;
    cerr << "      's' to save the current model," << endl;
    cerr << "      esc to quit." << endl;

    string cp("0");
    if (argc > 1) cp=argv[1];

    string imgpath("data\\persons");
    if (argc > 2) imgpath=argv[2];

    std::string cascade_path("haarcascade_frontalface_alt.xml");
    if (argc > 3) cascade_path=argv[3];

    if (argc == 1) {
        cerr << "please use : online [capture id or path] [img_path] [cascade_path]" << endl;
        cerr << "[current]  : online " << cp << " " << imgpath <<  " " << cascade_path << endl << endl;
        return 0;
    }

    namedWindow("reco");

    cv::CascadeClassifier cascade;
    bool clod = cascade.load(cascade_path);
    cerr << "cascade: " << clod  << endl;;

    VideoCapture cap;
    if (cp == "0") cap.open(0);
    else if (cp == "1") cap.open(1);
    else cap.open(cp);
    cerr << "capture(" << cp << ") : " << cap.isOpened() << endl;

    FaceRec reco;
    int n = reco.train(imgpath);
    cerr << "train: " << n << " classes." << endl;

    String save_model = "face.yml.gz";
    // alternatively, load a serialized model.
    //reco.load(save_model);

    vector<Mat> images;
    String caption = "";
    int showCaption = 0;
    int frameNo = 0;
    int state = NEUTRAL;
    Scalar color[3] = {
        Scalar(30,100,30),
        Scalar(10,10,160),
        Scalar(160,100,10),
    };

    while(cap.isOpened() && clod) {
        Mat frame;
        cap >> frame;
        if (frame.empty())
            break;

        // if it gets too slow, try with half the size.
        //pyrDown(frame,frame);

        if (state == PREDICT || state == CAPTURE) {
            Mat gray;
            cvtColor(frame, gray, COLOR_RGB2GRAY);

            std::vector<cv::Rect> faces;
            cascade.detectMultiScale(gray, faces, 1.2, 3,
                CASCADE_FIND_BIGGEST_OBJECT | CASCADE_DO_ROUGH_SEARCH  ,
                Size(40, 40), Size(300,300));

            if (faces.size() > 0) {
                Rect roi = faces[0];
                if ((state == CAPTURE) && (frameNo % 3 == 0)) {
                    Mat m;
                    resize(gray(roi), m, Size(FIXED_FACE,FIXED_FACE));
                    images.push_back(m);
                    cerr << ".";
                }
                if (state == PREDICT) {
                    caption = reco.predict(gray(roi));
                    if (caption != "") {
                        putText(frame, caption, Point(roi.x, roi.y+roi.width+13),
                            FONT_HERSHEY_PLAIN, 1.2, color[state], 1, LINE_AA);
                    }
                }
                rectangle(frame, roi, color[state]);
            }
        }
        for(int i=6,sc=6; i>1; i--,sc+=2) // status led
            circle(frame, Point(10,10), i, color[state]*(float(sc)/10), -1, LINE_AA);

        imshow("reco",frame);
        int k = waitKey(30);
        if (k == 27 ) break;
        if (k == 'p') state=PREDICT;
        if (k == 'c' && state != CAPTURE) {
            images.clear();
            state = CAPTURE;
        }
        if (k == 's') {
            cerr << "saved " << save_model << " : " << reco.save(save_model) << endl;
        }
        if (k =='n' || k==' ') {
            if ((state == CAPTURE) && (!images.empty())) {
                // query for a name, and write the images to that folder:
                cerr << endl << "please enter a name(empty to abort) :" << endl;
                char n[200];
                gets(n);
                if ((n[0] != 0) && (images.size() > 0)) {
                    String path = imgpath;
                    path += SEP;
                    path += n;
                    String cmdline = "mkdir ";
                    cmdline += path;
                    cerr << cmdline << endl;
                    system(cmdline.c_str()); // lame, but portable..
                    for (size_t i=0; i<images.size(); i++) {
                        imwrite(format("%s/%6d.png", path.c_str(), theRNG().next()), images[i]);
                    }
                    reco.train(imgpath);
                }
            }
            state = NEUTRAL;
        }
        frameNo++;
    }
    return 0;
}
